# Customer Sign-Up Analysis

Objective: Audit and analyse the customer-signups.csv dataset to identify data quality issues and derive business insights into user acquisition, plan selection, and marketing opt-in patterns

- Import Libraries and Load Dataset

In [105]:
import pandas as pd
import numpy as np

df=pd.read_csv(r"C:\Users\mymac\Downloads\customer_signups.csv")
df

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
0,CUST00000,Joshua Bryant,NaN,NaN,Instagram,NaN,basic,No,34,Female
1,CUST00001,Nicole Stewart,nicole1@example.com,02-01-24,LinkedIn,West,basic,Yes,29,Male
2,CUST00002,Rachel Allen,rachel2@example.com,03-01-24,Google,North,PREMIUM,Yes,34,Non-Binary
3,CUST00003,Zachary Sanchez,zachary3@mailhub.org,04-01-24,YouTube,NaN,Pro,No,40,Male
4,CUST00004,NaN,matthew4@mailhub.org,05-01-24,LinkedIn,West,Premium,No,25,Other
...,...,...,...,...,...,...,...,...,...,...
295,CUST00295,Gary Smith,gary95@example.com,22-10-24,Google,West,PREMIUM,Yes,40,NaN
296,CUST00296,Anthony Roberts,anthony96@mailhub.org,23-10-24,Google,Central,Basic,Yes,25,Female
297,CUST00297,Timothy Mclaughlin,NaN,24-10-24,Instagram,West,Basic,Yes,60,NaN
298,CUST00298,Justin Mcintyre,justin98@mailhub.org,25-10-24,YouTube,South,Premium,No,53,male


In [107]:
df.head()

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
0,CUST00000,Joshua Bryant,NaN,NaN,Instagram,NaN,basic,No,34,Female
1,CUST00001,Nicole Stewart,nicole1@example.com,02-01-24,LinkedIn,West,basic,Yes,29,Male
2,CUST00002,Rachel Allen,rachel2@example.com,03-01-24,Google,North,PREMIUM,Yes,34,Non-Binary
3,CUST00003,Zachary Sanchez,zachary3@mailhub.org,04-01-24,YouTube,NaN,Pro,No,40,Male
4,CUST00004,NaN,matthew4@mailhub.org,05-01-24,LinkedIn,West,Premium,No,25,Other


Identify Missing Values, Duplicates, and Data Types

In [108]:
# Check data types
print(df.dtypes)

customer_id         object
name                object
email               object
signup_date         object
source              object
region              object
plan_selected       object
marketing_opt_in    object
age                 object
gender              object
dtype: object


In [109]:
#Count missing values
print(df.isnull().sum())

customer_id          2
name                 9
email               34
signup_date          2
source               9
region              30
plan_selected        8
marketing_opt_in    10
age                 12
gender               8
dtype: int64


In [110]:
#Count duplicates by customer_id
duplicates = df.duplicated(subset='customer_id').sum()
print(f'Duplicates: {duplicates}')

Duplicates: 1


In [111]:
df['signup_date'] = pd.to_datetime(df['signup_date'], format='%d-%m-%y', errors='coerce')
df['signup_date'].head()
df['signup_date'].dtypes
df

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
0,CUST00000,Joshua Bryant,NaN,NaT,Instagram,NaN,basic,No,34,Female
1,CUST00001,Nicole Stewart,nicole1@example.com,2024-01-02,LinkedIn,West,basic,Yes,29,Male
2,CUST00002,Rachel Allen,rachel2@example.com,2024-01-03,Google,North,PREMIUM,Yes,34,Non-Binary
3,CUST00003,Zachary Sanchez,zachary3@mailhub.org,2024-01-04,YouTube,NaN,Pro,No,40,Male
4,CUST00004,NaN,matthew4@mailhub.org,2024-01-05,LinkedIn,West,Premium,No,25,Other
...,...,...,...,...,...,...,...,...,...,...
295,CUST00295,Gary Smith,gary95@example.com,2024-10-22,Google,West,PREMIUM,Yes,40,NaN
296,CUST00296,Anthony Roberts,anthony96@mailhub.org,2024-10-23,Google,Central,Basic,Yes,25,Female
297,CUST00297,Timothy Mclaughlin,NaN,2024-10-24,Instagram,West,Basic,Yes,60,NaN
298,CUST00298,Justin Mcintyre,justin98@mailhub.org,2024-10-25,YouTube,South,Premium,No,53,male


In [112]:
#Remove rows with missing or invalid customer_id and eliminate duplicates.

df = df[df['customer_id'].notna() & (df['customer_id'].str.strip() != '')]
df

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
0,CUST00000,Joshua Bryant,NaN,NaT,Instagram,NaN,basic,No,34,Female
1,CUST00001,Nicole Stewart,nicole1@example.com,2024-01-02,LinkedIn,West,basic,Yes,29,Male
2,CUST00002,Rachel Allen,rachel2@example.com,2024-01-03,Google,North,PREMIUM,Yes,34,Non-Binary
3,CUST00003,Zachary Sanchez,zachary3@mailhub.org,2024-01-04,YouTube,NaN,Pro,No,40,Male
4,CUST00004,NaN,matthew4@mailhub.org,2024-01-05,LinkedIn,West,Premium,No,25,Other
...,...,...,...,...,...,...,...,...,...,...
295,CUST00295,Gary Smith,gary95@example.com,2024-10-22,Google,West,PREMIUM,Yes,40,NaN
296,CUST00296,Anthony Roberts,anthony96@mailhub.org,2024-10-23,Google,Central,Basic,Yes,25,Female
297,CUST00297,Timothy Mclaughlin,NaN,2024-10-24,Instagram,West,Basic,Yes,60,NaN
298,CUST00298,Justin Mcintyre,justin98@mailhub.org,2024-10-25,YouTube,South,Premium,No,53,male


In [113]:
#Correct inconsistent plan names and unify text case for reliable analysis.
#PLAN_SELECTED

df.loc[:, 'plan_selected'] = df['plan_selected'].astype(str).str.strip().str.lower()
df.loc[:, 'plan_selected'] = df['plan_selected'].replace(['','unknownplan', ' ', 'none', 'nan'],'Unknown')
df.loc[:, 'plan_selected'] = df['plan_selected'].replace({
    'basic': 'Basic',
    'pro': 'Pro',
    'premium': 'Premium',
    'prem': 'Premium'
})

df['plan_selected'].value_counts()


plan_selected
Premium    99
Pro        93
Basic      92
Unknown    14
Name: count, dtype: int64

In [114]:
#Fix inconsistent entries (e.g., “M”, “Females”, “Nonbinary”) and replace blanks with “Unknown”.
#Gender
df.loc[:,'gender'] = df['gender'].astype(str).str.strip().str.title()


df.loc[df['gender'].str.isnumeric(), 'gender'] = 'Unknown' ##Replace invalid numeric values with 'Unknown'

#Fix inconsistent variations
df.loc[:,'gender'] = df['gender'].replace({
    'Females': 'Female',
    'Femail': 'Female',
    'F': 'Female',
    'M': 'Male',
    'Males': 'Male',
    'Non Binary': 'Non-Binary',
    'Nonbinary': 'Non-Binary',
    'Nb': 'Non-Binary',
    'Others': 'Other'
})

df.loc[:,'gender'] = df['gender'].replace(['', ' ', 'None', 'Nan', 'NaN'], 'Unknown') #Replace blanks or missing with 'Unknown'

df['gender'].value_counts() ##Check cleaned values


gender
Female        92
Male          91
Other         59
Non-Binary    42
Unknown       14
Name: count, dtype: int64

In [115]:
#Validate and Clean email Field
#email column:
#Strip spaces and replace blanks or None
df.loc[:,'email'] = df['email'].astype(str).str.strip()
df.loc[:,'email'] = df['email'].replace(['', ' ', 'None', 'nan', 'NaN'], np.nan)

#Validate email format using simple regex
import re
def validate_email(email):
    if pd.isna(email):
        return 'Unknown'
    pattern = r'^[\w\.-]+@[\w\.-]+\.\w+$'
    if re.match(pattern, email):
        return email
    else:
        return 'Unknown'

df.loc[:,'email'] = df['email'].apply(validate_email)

# Check results
df['email'].value_counts()


email
Unknown                    36
lisa11@mailhub.org          2
kyle99@inboxmail.net        1
patricia4@mailhub.org       1
kevin3@mailhub.org          1
                           ..
hannah6@example.com         1
cindy7@mailhub.org          1
margaret9@example.com       1
katherine10@example.com     1
mr.99@example.com           1
Name: count, Length: 262, dtype: int64

* Missing emails, regions, ages, names

* Duplicates (if any)

* Columns that need type conversion (like signup_date)

In [116]:
#Fix non-numeric or invalid age entries, convert to numeric type, and replace missing values with median age.
#Clean up Age column

df['age'] = pd.to_numeric(df['age'], errors='coerce')
df.loc[(df['age'] < 10) | (df['age'] > 100), 'age'] = np.nan
median_age = df['age'].median()
df.loc[:,'age'] = df['age'].fillna(median_age)             ##Fill missing ages with median
print(df['age'].dtype)

float64


C:\Users\mymac\AppData\Local\Temp\ipykernel_9592\2892184088.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age'] = pd.to_numeric(df['age'], errors='coerce')


In [117]:
#Region
df.loc[:,'region'] = df['region'].astype(str).str.strip()
df.loc[:,'region'] = df['region'].replace(['', ' ', 'nan', 'Na', 'None', 'N/A'],'Unknown')   #fill wiht Unknown

df['region'].value_counts()               #Check results


region
North      65
East       61
South      58
West       45
Central    39
Unknown    30
Name: count, dtype: int64

In [118]:
#Name 
df.loc[:,'name'] = df['name'].astype(str).str.strip()
df.loc[:,'name'] = df['name'].replace(['', ' ', 'None', 'nan', 'NaN'],'Unknown')

df['name'].value_counts()   #Check results

name
Unknown              9
Joshua Bryant        1
Kimberly Jackson     1
Brian Hicks          1
Patricia Powers      1
                    ..
Chelsey Rodriguez    1
Wayne Hopkins        1
Bryan George         1
Michelle Mcguire     1
Mr. Bruce Bridges    1
Name: count, Length: 290, dtype: int64

In [119]:
#Marketing_opt

df.loc[:,'marketing_opt_in'] = df['marketing_opt_in'].astype(str).str.strip()

df.loc[:,'marketing_opt_in'] = df['marketing_opt_in'].replace(['', ' ','Nil','NIL','None', 'none', 'nan', 'NaN'], 'Unknown')

df['marketing_opt_in'].value_counts() #check the unique values

marketing_opt_in
No         156
Yes        131
Unknown     11
Name: count, dtype: int64

In [120]:
## Source

df.loc[:,'source'] = df['source'].astype(str).str.strip().str.title()
df.loc[:,'source'] = df['source'].replace(['??', 'nan','Nan', 'Na', 'N/A', '', 'None'],'Unknown')  #Fill with 'Unknown'


df['source'].value_counts()             #Check values

source
Youtube      58
Google       50
Referral     49
Instagram    48
Facebook     40
Linkedin     38
Unknown      15
Name: count, dtype: int64

# Create Final Cleaned Dataset

Remove all rows containing any “Unknown” placeholders to create a high-quality dataset.

In [121]:
# Remove rows where any column has the string 'Unknown'
df_cleaned = df[~df.isin(['Unknown']).any(axis=1)]

# Check remaining rows
print(df_cleaned.shape)


(190, 10)


In [122]:
df_cleaned

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
1,CUST00001,Nicole Stewart,nicole1@example.com,2024-01-02,Linkedin,West,Basic,Yes,29.0,Male
2,CUST00002,Rachel Allen,rachel2@example.com,2024-01-03,Google,North,Premium,Yes,34.0,Non-Binary
5,CUST00005,John Gonzales,john5@mailhub.org,2024-01-06,Facebook,South,Premium,No,34.0,Other
7,CUST00007,Michael Bailey,michael7@mailhub.org,2024-01-08,Youtube,Central,Pro,Yes,60.0,Other
8,CUST00008,Bianca Morris,bianca8@example.com,2024-01-09,Referral,West,Pro,Yes,25.0,Male
...,...,...,...,...,...,...,...,...,...,...
291,CUST00291,Thomas Flores,thomas91@mailhub.org,2024-10-18,Youtube,South,Basic,No,60.0,Female
293,CUST00293,Yvonne Harding,yvonne93@inboxmail.net,2024-10-20,Google,Central,Pro,Yes,34.0,Male
294,CUST00294,Mrs. Jessica Smith,mrs.94@example.com,2024-10-21,Google,South,Pro,Yes,29.0,Other
296,CUST00296,Anthony Roberts,anthony96@mailhub.org,2024-10-23,Google,Central,Basic,Yes,25.0,Female


In [123]:
df_cleaned = df_cleaned.reset_index(drop=True)
df_cleaned 

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
0,CUST00001,Nicole Stewart,nicole1@example.com,2024-01-02,Linkedin,West,Basic,Yes,29.0,Male
1,CUST00002,Rachel Allen,rachel2@example.com,2024-01-03,Google,North,Premium,Yes,34.0,Non-Binary
2,CUST00005,John Gonzales,john5@mailhub.org,2024-01-06,Facebook,South,Premium,No,34.0,Other
3,CUST00007,Michael Bailey,michael7@mailhub.org,2024-01-08,Youtube,Central,Pro,Yes,60.0,Other
4,CUST00008,Bianca Morris,bianca8@example.com,2024-01-09,Referral,West,Pro,Yes,25.0,Male
...,...,...,...,...,...,...,...,...,...,...
185,CUST00291,Thomas Flores,thomas91@mailhub.org,2024-10-18,Youtube,South,Basic,No,60.0,Female
186,CUST00293,Yvonne Harding,yvonne93@inboxmail.net,2024-10-20,Google,Central,Pro,Yes,34.0,Male
187,CUST00294,Mrs. Jessica Smith,mrs.94@example.com,2024-10-21,Google,South,Pro,Yes,29.0,Other
188,CUST00296,Anthony Roberts,anthony96@mailhub.org,2024-10-23,Google,Central,Basic,Yes,25.0,Female


In [124]:
df = df_cleaned.copy()
df

,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
0,CUST00001,Nicole Stewart,nicole1@example.com,2024-01-02,Linkedin,West,Basic,Yes,29.0,Male
1,CUST00002,Rachel Allen,rachel2@example.com,2024-01-03,Google,North,Premium,Yes,34.0,Non-Binary
2,CUST00005,John Gonzales,john5@mailhub.org,2024-01-06,Facebook,South,Premium,No,34.0,Other
3,CUST00007,Michael Bailey,michael7@mailhub.org,2024-01-08,Youtube,Central,Pro,Yes,60.0,Other
4,CUST00008,Bianca Morris,bianca8@example.com,2024-01-09,Referral,West,Pro,Yes,25.0,Male
...,...,...,...,...,...,...,...,...,...,...
185,CUST00291,Thomas Flores,thomas91@mailhub.org,2024-10-18,Youtube,South,Basic,No,60.0,Female
186,CUST00293,Yvonne Harding,yvonne93@inboxmail.net,2024-10-20,Google,Central,Pro,Yes,34.0,Male
187,CUST00294,Mrs. Jessica Smith,mrs.94@example.com,2024-10-21,Google,South,Pro,Yes,29.0,Other
188,CUST00296,Anthony Roberts,anthony96@mailhub.org,2024-10-23,Google,Central,Basic,Yes,25.0,Female


Data Quality Summary

Review dataset structure post-cleaning & check counts, duplicates, and data completeness.

In [125]:

df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   customer_id       190 non-null    object        
 1   name              190 non-null    object        
 2   email             190 non-null    object        
 3   signup_date       185 non-null    datetime64[ns]
 4   source            190 non-null    object        
 5   region            190 non-null    object        
 6   plan_selected     190 non-null    object        
 7   marketing_opt_in  190 non-null    object        
 8   age               190 non-null    float64       
 9   gender            190 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(8)
memory usage: 15.0+ KB


,customer_id,name,email,signup_date,source,region,plan_selected,marketing_opt_in,age,gender
count,190,190,190,185,190,190,190,190,190.000000,190
unique,190,190,189,NaN,6,5,3,2,NaN,4
top,CUST00001,Nicole Stewart,lisa11@mailhub.org,NaN,Youtube,East,Basic,No,NaN,Female
freq,1,1,2,NaN,38,44,67,104,NaN,67
mean,NaN,NaN,NaN,2024-05-26 12:35:01.621621504,NaN,NaN,NaN,NaN,36.010526,NaN
min,NaN,NaN,NaN,2024-01-02 00:00:00,NaN,NaN,NaN,NaN,21.000000,NaN
25%,NaN,NaN,NaN,2024-03-11 00:00:00,NaN,NaN,NaN,NaN,26.000000,NaN
50%,NaN,NaN,NaN,2024-05-28 00:00:00,NaN,NaN,NaN,NaN,34.000000,NaN
75%,NaN,NaN,NaN,2024-08-08 00:00:00,NaN,NaN,NaN,NaN,40.000000,NaN
max,NaN,NaN,NaN,2024-10-25 00:00:00,NaN,NaN,NaN,NaN,60.000000,NaN


In [126]:
duplicates_all = df.duplicated()
print(f"Number of duplicate rows (all columns): {duplicates_all.sum()}")


Number of duplicate rows (all columns): 0


In [127]:
duplicates_id = df.duplicated(subset='customer_id')
print(f"Number of duplicate customer_ids: {duplicates_id.sum()}")


Number of duplicate customer_ids: 0


- Sign-Ups by Source, Region, and Plan

- Identify key acquisition channels and plan popularity across regions.

In [128]:
df['source'].value_counts()

source
Youtube      38
Instagram    33
Google       31
Referral     31
Facebook     29
Linkedin     28
Name: count, dtype: int64

In [129]:
df['region'].value_counts()

region
East       44
North      43
South      43
Central    31
West       29
Name: count, dtype: int64

In [130]:
df['plan_selected'].value_counts()

plan_selected
Basic      67
Premium    65
Pro        58
Name: count, dtype: int64

Marketing Opt-In by Gender

Assess differences in marketing opt-in behaviour across gender groups.



In [131]:
optin_gender = df.groupby(['gender', 'marketing_opt_in']).size().unstack(fill_value=0)
print(optin_gender)


marketing_opt_in  No  Yes
gender                   
Female            33   34
Male              37   23
Non-Binary        12   12
Other             22   17


Age Summary Statistics
Understand demographic spread of the customer base.




In [132]:
df['age'].describe()

count    190.000000
mean      36.010526
std       11.151193
min       21.000000
25%       26.000000
50%       34.000000
75%       40.000000
max       60.000000
Name: age, dtype: float64

# Business Questions

In [133]:
#Q.1 Which acquisition source brought in the most users last month?

# Find last month in the data
last_month = df['signup_date'].max().month

# Filter data for that month
last_month_data = df[df['signup_date'].dt.month == last_month]

# Count sign-ups by source
print(last_month_data['source'].value_counts())
print("Last month in dataset:", df['signup_date'].max().strftime('%B %Y'))



source
Google       5
Referral     3
Youtube      3
Instagram    2
Linkedin     1
Facebook     1
Name: count, dtype: int64
Last month in dataset: October 2024


In [134]:
#Q.2 Are older users more or less likely to opt in to marketing?
#Average opt-in rate by age

print(df.groupby('marketing_opt_in')['age'].mean())



marketing_opt_in
No     36.057692
Yes    35.953488
Name: age, dtype: float64


In [135]:
#Q.3 Which plan is most commonly selected, and by which age group?
print(df['plan_selected'].value_counts())
print(df.groupby('plan_selected')['age'].mean())


plan_selected
Basic      67
Premium    65
Pro        58
Name: count, dtype: int64
plan_selected
Basic      34.985075
Premium    37.200000
Pro        35.862069
Name: age, dtype: float64


# Final Summary


Summarise overall findings:

* Data cleaned and standardised (190 valid records retained).

* Most users signed up via YouTube and Google.

* “Basic” plan is most popular.

* Marketing opt-in similar across age groups.

* West region has lowest data completeness.